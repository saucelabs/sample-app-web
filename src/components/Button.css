.btn {
  box-sizing: border-box;
  cursor: pointer;
  outline: none;
  padding: 6px 28px;
  font-size: 14px;
  font-family: "Roboto", Arial, Helvetica, sans-serif;
  display: inline-block;
  text-decoration: none;
  text-transform: uppercase;
}

.btn_primary {
  border: 2px solid #e2231a;
  color: #e2231a;
  background-color: #fff;
}

.btn_secondary {
  border: 2px solid #474c55;
  background-color: #fff;
  color: #474c55;
}

.btn_action {
  background-color: #e2231a;
  border: 2px solid #e2231a;
  color: #fff;
}

.btn_secondary.back {
  position: relative;
}

.btn_secondary .back-image {
  position: absolute;
  left: 10px;
  top: 9px;
  height: 11px;
  width: 11px;
}

.btn_small {
  width: 160px;
}

.btn_medium {
  width: 220px;
}

.btn_large {
  width: 100%;
}

@media only screen and (max-width: 900px) {
  .btn_small,
  .btn_medium,
  .btn_large {
    width: 100%;
  }
}
