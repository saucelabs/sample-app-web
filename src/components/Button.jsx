import React from "react";
import PropTypes from "prop-types";
import "./Button.css";
import backPng from "../assets/img/arrow.png";

export const BUTTON_TYPES = {
  ACTION: "action",
  BACK: "secondary back",
  PRIMARY: "primary",
  SECONDARY: "secondary",
};
export const BUTTON_SIZES = {
  SMALL: "small",
  MEDIUM: "medium",
  LARGE: "large",
};
const Button = ({
  customClass,
  label,
  onClick,
  size,
  testId,
  type,
  ...props
}) => {
  const buttonTypeClass = ` btn_${type}`;
  const extraClass = customClass ? ` ${customClass}` : "";
  const buttonSize = ` btn_${size}`;
  /* istanbul ignore next */
  const BackImage = () => (
    <img src={backPng} className="back-image" alt="Go back" />
  );

  return (
    <button
      className={`btn${buttonTypeClass}${buttonSize}${extraClass}`}
      {...(testId
        ? {
            "data-test": testId,
            id: testId,
            name: testId,
          }
        : {})}
      onClick={onClick}
      {...props}
    >
      {type === BUTTON_TYPES.BACK && <BackImage />}
      {label}
    </button>
  );
};

Button.propTypes = {
  /**
   * A custom class
   */
  customClass: PropTypes.string,
  /**
   * The label
   */
  label: PropTypes.string.isRequired,
  /**
   * The on click handler
   */
  onClick: PropTypes.func.isRequired,
  /**
   * Size of the button
   */
  size: PropTypes.oneOf(Object.values(BUTTON_SIZES)),
  /**
   * The test id
   */
  testId: PropTypes.string,
  /**
   * What type of field is it
   */
  type: PropTypes.oneOf(Object.values(BUTTON_TYPES)),
};

Button.defaultProps = {
  customClass: undefined,
  size: BUTTON_SIZES.LARGE,
  testId: undefined,
  type: BUTTON_TYPES.PRIMARY,
};

export default Button;
