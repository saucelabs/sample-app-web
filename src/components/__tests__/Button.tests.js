import React from "react";
import { shallow } from "enzyme";
import Button, { BUTTON_SIZES, BUTTON_TYPES } from "../Button";

let props;

describe("Button", () => {
  beforeEach(() => {
    props = {
      label: "Default button",
      onClick: () => {},
    };
  });

  it("should render correctly with the required options", () => {
    const component = shallow(<Button {...props} />);

    expect(component).toMatchSnapshot();
  });

  it("should render correctly with a custom class", () => {
    const component = shallow(<Button customClass="custom-class" {...props} />);

    expect(component).toMatchSnapshot();
  });

  it("should render a Action button", () => {
    const component = shallow(<Button type={BUTTON_TYPES.ACTION} {...props} />);

    expect(component).toMatchSnapshot();
  });

  it("should render a Back button", () => {
    const component = shallow(<Button type={BUTTON_TYPES.BACK} {...props} />);

    expect(component).toMatchSnapshot();
  });

  it("should render a Primary button", () => {
    const component = shallow(
      <Button type={BUTTON_TYPES.PRIMARY} {...props} />
    );

    expect(component).toMatchSnapshot();
  });

  it("should render a Secondary button", () => {
    const component = shallow(
      <Button type={BUTTON_TYPES.SECONDARY} {...props} />
    );

    expect(component).toMatchSnapshot();
  });

  it("should render a Small button", () => {
    const component = shallow(<Button size={BUTTON_SIZES.SMALL} {...props} />);

    expect(component).toMatchSnapshot();
  });

  it("should render a Medium button", () => {
    const component = shallow(<Button size={BUTTON_SIZES.MEDIUM} {...props} />);

    expect(component).toMatchSnapshot();
  });

  it("should render a Large button", () => {
    const component = shallow(<Button size={BUTTON_SIZES.LARGE} {...props} />);

    expect(component).toMatchSnapshot();
  });

  it("should render a button with a testId", () => {
    const component = shallow(<Button testId="test-id" {...props} />);

    expect(component).toMatchSnapshot();
  });

  it("should render a button with custom properties", () => {
    const component = shallow(<Button foo="bar" bar="foo" {...props} />);

    expect(component).toMatchSnapshot();
  });
});
