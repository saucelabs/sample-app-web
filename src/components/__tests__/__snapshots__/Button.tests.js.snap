// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render a Action button 1`] = `
<button
  className="btn btn_action btn_large"
  onClick={[Function]}
>
  Default button
</button>
`;

exports[`Button should render a Back button 1`] = `
<button
  className="btn btn_secondary back btn_large"
  onClick={[Function]}
>
  <BackImage />
  Default button
</button>
`;

exports[`Button should render a Large button 1`] = `
<button
  className="btn btn_primary btn_large"
  onClick={[Function]}
>
  Default button
</button>
`;

exports[`Button should render a Medium button 1`] = `
<button
  className="btn btn_primary btn_medium"
  onClick={[Function]}
>
  Default button
</button>
`;

exports[`Button should render a Primary button 1`] = `
<button
  className="btn btn_primary btn_large"
  onClick={[Function]}
>
  Default button
</button>
`;

exports[`Button should render a Secondary button 1`] = `
<button
  className="btn btn_secondary btn_large"
  onClick={[Function]}
>
  Default button
</button>
`;

exports[`Button should render a Small button 1`] = `
<button
  className="btn btn_primary btn_small"
  onClick={[Function]}
>
  Default button
</button>
`;

exports[`Button should render a button with a testId 1`] = `
<button
  className="btn btn_primary btn_large"
  data-test="test-id"
  id="test-id"
  name="test-id"
  onClick={[Function]}
>
  Default button
</button>
`;

exports[`Button should render a button with custom properties 1`] = `
<button
  bar="foo"
  className="btn btn_primary btn_large"
  foo="bar"
  onClick={[Function]}
>
  Default button
</button>
`;

exports[`Button should render correctly with a custom class 1`] = `
<button
  className="btn btn_primary btn_large custom-class"
  onClick={[Function]}
>
  Default button
</button>
`;

exports[`Button should render correctly with the required options 1`] = `
<button
  className="btn btn_primary btn_large"
  onClick={[Function]}
>
  Default button
</button>
`;
